xsnap/documentation/Compartments.md

```diff
@@ -0,0 +1,618 @@
+	
+	const c2 = new Compartment({}, moduleMap, {});
+	const fooNS2 = await c2.import("foo");
+	const barNS2 = await c2.import("bar");
+	
+	print(getterCount); // 2;
+	print(setterCount); // 0;
+	print(fooNS1.default()); // 0;
+	print(barNS1.default()); // 0;
+	print(fooNS2.default()); // 0;
+	print(barNS2.default()); // 0;
+
+#### module specifiers
+
+When a module map property value is a module specifier, the module specifier is used by the parent compartment to load and initialize a new module into the compartment.
+
+	const uri = import.meta.uri;
+	const foo = uri.slice(0, uri.lastIndexOf("/") + 1) + "moduleMap-fixture.js";
+	
+	let getterCount = 0;
+	let setterCount = 0;
+	const moduleMap = {
+		foo,
+		get bar() {
+			getterCount++;
+			return this.foo;
+		},
+		set bar(it) {
+			setterCount++;
+			this.foo = it;
+		},
+	};
+	
+	const c1 = new Compartment({}, moduleMap, {});
+	const fooNS1 = await c1.import("foo");
+	const barNS1 = await c1.import("bar");
+	
+	const c2 = new Compartment({}, moduleMap, {});
+	const fooNS2 = await c2.import("foo");
+	const barNS2 = await c2.import("bar");
+	
+	print(getterCount); // 2;
+	print(setterCount); // 0;
+	print(fooNS1.default()); // 0;
+	print(barNS1.default()); // 0;
+	print(fooNS2.default()); // 0;
+	print(barNS2.default()); // 0;
+
+
+### options.globalLexicals
+
+The `globalLexicals` option is an object that initializes the global lexical scope of the compartment:
+
+- property names become variable or constant names,
+- property values become variable or constant values. 
+
+If the property is writable, a variable (`let`) is created, else a constant (`const`) is created.
+
+	let getterCount = 0;
+	let setterCount = 0;
+	const globalLexicals = {
+		foo: 0,
+		get bar() {
+			getterCount++;
+			return this.foo;
+		},
+		set bar(it) {
+			setterCount++;
+			this.foo = it;
+		},
+		shared: {
+			foo: 0,
+			get bar() {
+				return this.foo;
+			},
+			set bar(it) {
+				this.foo = it;
+			},
+		}
+	};
+	
+	const c1 = new Compartment({ print }, {}, { globalLexicals});
+	c1.evaluate(`
+		foo++;
+		bar++;
+		shared.foo++;
+		shared.bar++;
+	`);
+	
+	const c2 = new Compartment({ print }, {}, { globalLexicals });
+	c2.evaluate(`
+		foo++;
+		bar++;
+		shared.foo++;
+		shared.bar++;
+	`);
+	
+	print(getterCount); // 2;
+	print(setterCount); // 0;
+	print(globalLexicals.foo); // 0;
+	print(globalLexicals.bar); // 0;
+	print(globalLexicals.shared.foo); // 4;
+	print(globalLexicals.shared.bar); // 4;
+	c1.evaluate(`
+		print(foo); // 1;
+		print(bar); // 1;
+	`);
+	print(c1.globalThis.foo); // undefined;
+	print(c1.globalThis.bar); // undefined;
+	c2.evaluate(`
+		print(foo); // 1;
+		print(bar); // 1;
+	`);
+	print(c2.globalThis.foo); // undefined;
+	print(c2.globalThis.bar); // undefined;
+
+### options.moduleMapHook(specifier)
+
+The `moduleMapHook` option is a function that takes a module specifier and returns a module namespace, a module descriptor, a module specifier or `undefined`.
```

Comment:
Have module descriptors fully replaced `StaticModuleRecord` entirely? I’m expecting module fragment syntax to produce a static module record that is distinguishable by a brand check, and that the static module record constructor might still be useful in the presence of module descriptors to amortize the cost of compiling.

Reply:
I like that module descriptors create a clear path for additional properties `meta` and `specifier`. Perhaps `record` could be the distinguished property for a module descriptor that contains a reusable, precompiled static module record.